File: app/src/main/java/com/zxcx/zhizhe/ui/my/feedback/feedback/FeedbackActivity.java
package com.zxcx.zhizhe.ui.my.feedback.feedback;

import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.widget.EditText;
import android.widget.TextView;

import com.meituan.android.walle.WalleChannelReader;
import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.event.CancelFeedbackConfirmEvent;
import com.zxcx.zhizhe.mvpBase.MvpActivity;
import com.zxcx.zhizhe.utils.Constants;
import com.zxcx.zhizhe.utils.Utils;
import com.zxcx.zhizhe.widget.CancelFeedbackConfirmDialog;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.OnClick;

public class FeedbackActivity extends MvpActivity<FeedbackPresenter> implements FeedbackContract.View {


    @BindView(R.id.tv_toolbar_right)
    TextView mTvToolbarRight;
    @BindView(R.id.et_feedback_content)
    EditText mEtFeedbackContent;
    @BindView(R.id.tv_feedback_residue)
    TextView mTvFeedbackResidue;
    @BindView(R.id.et_feedback_phone)
    EditText mEtFeedbackPhone;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_feedback);
        ButterKnife.bind(this);
        EventBus.getDefault().register(this);

        initToolBar(R.string.tv_my_feedback);
        mTvToolbarRight.setText("提交");
        mTvToolbarRight.setTextColor(ContextCompat.getColor(mActivity,R.color.color_text_enable_blue));
    }

    @Override
    public void onBackPressed() {
        if (mEtFeedbackContent.length() > 0) {
            CancelFeedbackConfirmDialog dialog = new CancelFeedbackConfirmDialog();
            dialog.show(getFragmentManager(), "");
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    @Override
    protected FeedbackPresenter createPresenter() {
        return new FeedbackPresenter(this);
    }

    @Override
    public void postSuccess() {
        finish();
    }

    @Override
    public void postFail(String msg) {
        toastShow(msg);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(CancelFeedbackConfirmEvent event) {
        finish();
    }

    @OnClick(R.id.tv_toolbar_right)
    public void onMTvToolbarRightClicked() {
        String content = mEtFeedbackContent.getText().toString();
        String contact = mEtFeedbackPhone.getText().toString();
        int appType = Constants.APP_TYPE;
        String appChannel = WalleChannelReader.getChannel(this);
        String appVersion = Utils.getAppVersionName(this);
        mPresenter.feedback(content, contact, appType, appChannel, appVersion);
    }

    @OnCheckedChanged(R.id.et_feedback_content)
    public void onMEtFeedbackContentChanged() {
        if (mEtFeedbackContent.length() > 0) {
            mTvToolbarRight.setEnabled(true);
        } else {
            mTvToolbarRight.setEnabled(false);
        }
        mTvFeedbackResidue.setText(String.valueOf(199-mEtFeedbackContent.length()));
    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/my/feedback/feedback/FeedbackActivity.kt
package com.zxcx.zhizhe.ui.my.feedback.feedback

import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import com.meituan.android.walle.WalleChannelReader
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.mvpBase.MvpActivity
import com.zxcx.zhizhe.utils.Constants
import com.zxcx.zhizhe.utils.Utils
import com.zxcx.zhizhe.utils.afterTextChanged
import com.zxcx.zhizhe.utils.getColorForKotlin
import kotlinx.android.synthetic.main.activity_feedback.*
import kotlinx.android.synthetic.main.toolbar.*

/**
 * 反馈页面
 */

class FeedbackActivity : MvpActivity<FeedbackPresenter>(), FeedbackContract.View {

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_feedback)

		initToolBar()
		tv_toolbar_right.visibility = View.VISIBLE
		tv_toolbar_right.text = "完成"
		tv_toolbar_right.isEnabled = false
		tv_toolbar_right.setTextColor(ContextCompat.getColorStateList(mActivity, R.color.color_text_enable_blue))
	}

	override fun onDestroy() {
		Utils.closeInputMethod(this)
		super.onDestroy()
	}

	override fun createPresenter(): FeedbackPresenter {
		return FeedbackPresenter(this)
	}

	override fun postSuccess() {
		toastShow("反馈已提交，感谢你的支持")
		finish()
	}

	override fun postFail(msg: String) {
		toastShow(msg)
	}

	override fun setListener() {

		tv_toolbar_right.setOnClickListener {
			val content = et_feedback_content.text.toString()
			val appType = Constants.APP_TYPE
			val appChannel = WalleChannelReader.getChannel(this)
			val appVersion = Utils.getAppVersionName(this)
			mPresenter.feedback(content, appType, appChannel, appVersion)
		}

		et_feedback_content.afterTextChanged {
			tv_toolbar_right.isEnabled = it.length >= 10
			tv_feedback_residue.text = getString(R.string.tv_feedback_residue, it.length, 600 - it.length)
			if (it.length == 600) {
				tv_feedback_residue.setTextColor(mActivity.getColorForKotlin(R.color.red))
			} else {
				tv_feedback_residue.setTextColor(mActivity.getColorForKotlin(R.color.text_color_d2))
			}
		}

		et_feedback_content.setOnEditorActionListener { v, actionId, event ->
			//此处会响应2次 分别为ACTION_DOWN和ACTION_UP
			if (actionId == EditorInfo.IME_ACTION_SEARCH
					|| actionId == EditorInfo.IME_ACTION_DONE
					|| event != null && KeyEvent.KEYCODE_ENTER == event.keyCode && KeyEvent.ACTION_DOWN == event.action) {

				Utils.closeInputMethod(et_feedback_content)

				return@setOnEditorActionListener true
			}
			return@setOnEditorActionListener false
		}

	}
}
